### ISSUE TITLE
Bug Bounty Report: Leakage of Heroku API Key from JavaScript Files

### ISSUE THREAD

**Issue by h4ckersanjith** [runbox/runbox7#1636]
Summary
An API key associated with a Heroku account has been exposed in publicly accessible JavaScript files on the website. This exposure occurs because the JavaScript file containing sensitive information was not adequately secured or filtered, allowing unauthorized parties to extract the API key. If misused, this key could lead to unauthorized access to Heroku services, potentially compromising the security of the associated applications.

Vulnerability Details
Vulnerable Component: JavaScript file(s) served from the website.

Vulnerability Type: Sensitive Data Exposure (API Key Leakage).

Description: During inspection of the public-facing JavaScript files (accessible at https://<website-url>/path/to/js-file.js), I discovered that the Heroku API key was hardcoded in the code. This key provides unauthorized access to sensitive resources or services within Heroku and can be misused if retrieved by malicious actors.

Affected URL(s):

https://runbox.com/_js/jquery-1.5.1.min.js?v=20201006_074600

Example Exposed Key:

javascript
Copy code
const herokuApiKey = "your-heroku-api-key-here";
The API key is visible directly in the source code of the file and can be easily retrieved by anyone with access to the public website.

Steps to Reproduce
Open a web browser and navigate to the URL where the JavaScript file is located: https://runbox.com/_js/jquery-1.5.1.min.js?v=20201006_074600

View the file’s content by right-clicking on the page and selecting "View Page Source" or using browser developer tools (Inspect → Network → JavaScript file).

LEAKED CREDENTIALS:
Heroku API KEY  ->      D27CDB6E-AE6D-11cf-96B8-444553540000

Search for the string that matches the Heroku API key or use the key value in the code to identify it.
The API key is present as part of the JavaScript code and can be copied or retrieved from the file.
Impact
The exposure of the Heroku API key can have significant consequences:

Unauthorized Access: Any attacker can use the exposed API key to interact with the Heroku platform, including deploying or modifying applications, accessing sensitive environment variables, and interacting with connected resources.
Data Breach: If the attacker has sufficient permissions, they could compromise sensitive application data, database configurations, or perform other malicious actions.
Service Disruption: The exposed API key could allow malicious users to initiate unauthorized actions, such as scaling up or down resources, leading to potential service disruptions.
Suggested Remediation
Remove API Key from Public Code: The API key should never be hardcoded directly into public-facing JavaScript files. Instead, it should be kept in environment variables or secure storage mechanisms.
Use Environment Variables: Move the Heroku API key to a backend server environment, which is not exposed to the client side. Only the server should have access to the API key, and it should not be accessible via the frontend code.
Use .env Files or Secret Management Solutions: For local development, use .env files and for production, use Heroku’s built-in environment variable management system or a secret management service like Vault or AWS Secrets Manager to securely store and manage sensitive keys.
Audit All JavaScript Files: Perform a thorough audit of all publicly accessible JavaScript files to ensure that no sensitive information, such as API keys or credentials, is exposed.
Regenerate API Key: After securing the key, regenerate it in the Heroku dashboard to prevent any misuse of the previously exposed key.
References
Heroku: Managing API Keys
OWASP: API Key Management Best Practices
Conclusion
The exposed Heroku API key in the JavaScript file is a significant security risk that could lead to unauthorized access to the associated Heroku account. Immediate action should be taken to secure the API key, regenerate it, and remove any sensitive data from public-facing code. By following the suggested remediation steps, the risk of exploitation can be mitigated.

Thanks & Regards
Sanjith Roshan U 
Security Researcher

**Comment by gtandersen** [comment]
@h4ckersanjith Thanks for this and the related bug reports. However, we are unable to find the strings mentioned in the code.

**Comment by h4ckersanjith** [comment]
Any problem?

**Comment by gtandersen** [comment]
@h4ckersanjith For instance in this bug report the string `Heroku` does not appear anywhere in the jquery javascript, nor anywhere else in our _js/ files. We get the same results for the other, similar bug reports.

**Comment by h4ckersanjith** [comment]
You can't reproduce the issue right?

**Comment by gtandersen** [comment]
Correct.

**Comment by h4ckersanjith** [comment]
So what's the solution now.
But it is very important to fix

**Comment by castaway** [comment]
Hi,

I looked at the code of the SecretFinder tool, it is using this regular expression to find heroku keys, for example: `[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}` 
 see https://github.com/m4ll0k/SecretFinder/blob/master/SecretFinder.py#L72
 
This potential key matches any GUID type string in the codebase.. in this instance I found it was matching a part of jquery that relates to shockwave flash, and parts of our TinyMCE install: 
 ```
         writeEmbed(
                'D27CDB6E-AE6D-11cf-96B8-444553540000',
                'http://download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=6,0,40,0',
                'application/x-shockwave-flash',
                p
        );
```
If you google this key you get: `{D27CDB6E-AE6D-11CF-96B8-444553540000} , Microsoft's GUID for representing Shockwave Flash` so it is a GUID, which is not a Heroku API key.

The "twilio account sid" items that you listed in the VPN client javascript, are sub parts of an embedded font, which looks like `src: 
url('data:application/font-woff;base64,d09GRgABAAAAAArUAA8AAAAAEzQAAQAAAAAAAAAA ....` and contains substrings that begin with "AC", these are not Twilio Account sids.

We do appreciate the bug hunting, and will check all reports carefully, however these particular ones are not bugs, they are normal working code.
